src/core/defs.jl

```diff
@@ -187,6 +187,24 @@ dim_value_dict(md::ModelDef) = Dict([name => collect(values(dim)) for (name, dim
 Base.haskey(md::ModelDef, name::Symbol) = haskey(md.dimensions, name)
 
 isuniform(md::ModelDef) = md.is_uniform
+
+
+# Helper function invoked when the user resets the time dimension with set_dimension!
+# This function calls set_run_period! on each component definition to reset the first and last values.
+function reset_run_periods!(md, first, last)
+
+    for comp_def in values(md.comp_defs)
+        if comp_def.first < first 
```

Comment:
I recommend using a functional style when possible to avoid spreading implementation details of types throughout the code. So instead of directly accessing `comp_def.first`, I'd use these functions (already defined):
```
first_period(comp_def::ComponentDef) = comp_def.first
last_period(comp_def::ComponentDef) = comp_def.last
```

The complier will inline these, so it's the same at run-time, but hiding struct details behind a functional interface allows more flexibility if we change the structure at some point.